aiidalab/home.py

```diff
@@ -1,37 +1,99 @@
 """Module to generate AiiDA lab home page."""
 from os import path
 from glob import glob
+from functools import wraps
 
 import json
+import traitlets
 import ipywidgets as ipw
 from IPython.display import display
 
 # AiiDA lab imports.
 from .app import AiidaLabApp
 from .config import AIIDALAB_APPS
 from .utils import load_widget, load_app_registry
+from .widgets import UpdateAvailableInfoWidget
 
 
-def mk_buttons(name):
-    """Make buttons to move the app up or down."""
+def create_app_widget_move_buttons(name):
+    """Make buttons to move the app widget up or down."""
     layout = ipw.Layout(width="40px")
-    btn_box = ipw.HTML("""
+    app_widget_move_buttons = ipw.HTML("""
     <a href=./start.ipynb?move_up={name} title="Move it up"><i class='fa fa-arrow-up' style='color:#337ab7;font-size:2em;' ></i></a>
     <a href=./start.ipynb?move_down={name} title="Move it down"><i class='fa fa-arrow-down' style='color:#337ab7;font-size:2em;' ></i></a>
     """.format(name=name),
-                       layout=layout)
-    btn_box.layout.margin = "50px 0px 0px 0px"
+                                       layout=layout)
+    app_widget_move_buttons.layout.margin = "50px 0px 0px 0px"
 
-    return btn_box
+    return app_widget_move_buttons
+
+
+def _workaround_property_lock_issue(func):
+    """Work-around for issue with the ipw.Accordion widget.
+
+    The widget does not report changes to the .selected_index trait when displayed
+    within a custom ipw.Output instance. However, the change is somewhat cryptic reported
+    by a change to the private '_property_lock' trait. We observe changes to that trait
+    and convert the change argument into a form that is more like the one expected by
+    downstream handlers.
+    """
+
+    @wraps(func)
+    def _inner(self, change):
+        if change['name'] == '_property_lock':
+            if 'selected_index' in change['new']:
+                fixed_change = change.copy()
+                fixed_change['name'] = 'selected_index'
+                fixed_change['new'] = change['new']['selected_index']
+                del fixed_change['old']
+                return func(self, fixed_change)
+
+        return func(self, change)
+
+    return _inner
 
 
 class AiidaLabHome:
     """Class that mananges the appearance of the AiiDA lab home page."""
 
+    class _AppWidgets(dict):
```

Comment:
can you provide an example of a problem you try to solve with this?

Reply:
It enables https://github.com/aiidalab/aiidalab/blob/a622362bf20d4867f62c5e7383f7b10c2e2cbfc5/aiidalab/home.py#L113

It means that app widgets are created when needed and then cached within this dictionary. I can refactor the code to make that more explicit if this is difficult to follow.